quarkus.rest.jackson.optimization.enable-reflection-free-serializers=true fails native build of StartStopTS microprofile application

### Describe the bug

As suggested in https://github.com/quarkusio/quarkus/issues/42506#issuecomment-2311076936 I tried to run RSS check with `quarkus.rest.jackson.optimization.enable-reflection-free-serializers=true` added to the sample microprofile application (https://github.com/rsvoboda/quarkus-startstop/tree/reactive-jackson-and-health/app-full-microprofile).

Unfortunately, native image build fails when this property is added. I'm using `mvn -f app-full-microprofile clean package -Dnative -Dquarkus.version=999-SNAPSHOT -Dquarkus.rest.jackson.optimization.enable-reflection-free-serializers=true` command.

Native compilation has issue with https://github.com/rsvoboda/quarkus-startstop/blob/reactive-jackson-and-health/app-full-microprofile/src/main/java/com/example/quarkus/serialization/dto/NestedInterface.java. Native compilation without enable-reflection-free-serializers property works well.

Message from the build:
`Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Discovered unresolved method during parsing: com.example.quarkus.serialization.dto.NestedInterface.getCharacter(). This error is reported at image build time because class com.example.quarkus.serialization.dto.NestedInterface$quarkusjacksonserializer is registered for linking at image build time by command line and command line.`


Detailed stack trace:
```
[2/8] Performing analysis...  [08:41:30,400 WARN  [io.net.res.dns.DnsServerAddressStreamProviders] Can not find io.netty.resolver.dns.macos.MacOSDnsServerAddressStreamProvider in the classpath, fallback to system defaults. This may result in incorrect DNS resolutions on MacOS. Check whether you have a dependency on 'io.netty:netty-resolver-dns-native-macos'
*****]                                                                   (11.9s @ 1.00GB)
   12,391 reachable types   (87.2% of   14,208 total)
   17,617 reachable fields  (57.0% of   30,904 total)
   63,763 reachable methods (57.3% of  111,277 total)
    3,907 types,   140 fields, and 3,314 methods registered for reflection
       62 types,    67 fields, and    55 methods registered for JNI access
        5 native libraries: -framework CoreServices, -framework Foundation, dl, pthread, z

Error: Discovered unresolved method during parsing: com.example.quarkus.serialization.dto.NestedInterface.getCharacter(). This error is reported at image build time because class com.example.quarkus.serialization.dto.NestedInterface$quarkusjacksonserializer is registered for linking at image build time by command line and command line.
Error encountered while parsing com.example.quarkus.serialization.dto.NestedInterface$quarkusjacksonserializer.serialize(Unknown Source)
Parsing context:
   at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:490)
   at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:334)
   at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4799)
   at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4040)
   at io.vertx.core.json.jackson.DatabindCodec.toString(DatabindCodec.java:163)
   at io.vertx.core.json.JsonObject.encode(JsonObject.java:1034)
   at io.vertx.core.json.JsonObject.toString(JsonObject.java:1158)
   at root method.(Unknown Source)

Detailed message:

com.oracle.svm.core.util.UserError$UserException: Discovered unresolved method during parsing: com.example.quarkus.serialization.dto.NestedInterface.getCharacter(). This error is reported at image build time because class com.example.quarkus.serialization.dto.NestedInterface$quarkusjacksonserializer is registered for linking at image build time by command line and command line.
Error encountered while parsing com.example.quarkus.serialization.dto.NestedInterface$quarkusjacksonserializer.serialize(Unknown Source)
Parsing context:
   at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:490)
   at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:334)
   at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4799)
   at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4040)
   at io.vertx.core.json.jackson.DatabindCodec.toString(DatabindCodec.java:163)
   at io.vertx.core.json.JsonObject.encode(JsonObject.java:1034)
   at io.vertx.core.json.JsonObject.toString(JsonObject.java:1158)
   at root method.(Unknown Source)

Detailed message:

	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:85)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:248)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:814)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:592)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:550)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:539)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:721)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.start(NativeImageGeneratorRunner.java:143)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:98)
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Discovered unresolved method during parsing: com.example.quarkus.serialization.dto.NestedInterface.getCharacter(). This error is reported at image build time because class com.example.quarkus.serialization.dto.NestedInterface$quarkusjacksonserializer is registered for linking at image build time by command line and command line.
Error encountered while parsing com.example.quarkus.serialization.dto.NestedInterface$quarkusjacksonserializer.serialize(Unknown Source)
Parsing context:
   at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:490)
   at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:334)
   at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4799)
   at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4040)
   at io.vertx.core.json.jackson.DatabindCodec.toString(DatabindCodec.java:163)
   at io.vertx.core.json.JsonObject.encode(JsonObject.java:1034)
   at io.vertx.core.json.JsonObject.toString(JsonObject.java:1158)
   at root method.(Unknown Source)

Detailed message:

	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:126)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:809)
	... 6 more
Caused by: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: com.example.quarkus.serialization.dto.NestedInterface.getCharacter(). This error is reported at image build time because class com.example.quarkus.serialization.dto.NestedInterface$quarkusjacksonserializer is registered for linking at image build time by command line and command line.
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.reportUnresolvedElement(SharedGraphBuilderPhase.java:550)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.reportUnresolvedElement(SharedGraphBuilderPhase.java:544)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedMethod(SharedGraphBuilderPhase.java:471)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedInvoke(SharedGraphBuilderPhase.java:368)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genInvokeVirtual(BytecodeParser.java:1740)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5439)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3431)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.iterateBytecodesForBlock(SharedGraphBuilderPhase.java:743)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.handleBytecodeBlock(BytecodeParser.java:3391)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3233)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1137)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.build(SharedGraphBuilderPhase.java:162)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:1029)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:101)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:116)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:434)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.AnalysisParsedGraph.parseBytecode(AnalysisParsedGraph.java:146)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisMethod.parseGraph(AnalysisMethod.java:895)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisMethod.ensureGraphParsedHelper(AnalysisMethod.java:860)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisMethod.ensureGraphParsed(AnalysisMethod.java:843)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:186)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:621)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.createFlowsGraph(MethodTypeFlow.java:167)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureFlowsGraphCreated(MethodTypeFlow.java:153)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.getOrCreateMethodFlowsGraphInfo(MethodTypeFlow.java:111)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.typestate.DefaultVirtualInvokeTypeFlow.onObservedUpdate(DefaultVirtualInvokeTypeFlow.java:114)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:620)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.PointsToAnalysis$1.run(PointsToAnalysis.java:491)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:187)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:171)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

### Expected behavior

Native image get properly produced.

### Actual behavior

Native image build fails

### How to Reproduce?

```
sdk use java 21.0.2-graalce
git clone https://github.com/rsvoboda/quarkus-startstop -b reactive-jackson-and-health
mvn -f app-full-microprofile clean package -Dnative -Dquarkus.version=999-SNAPSHOT -Dquarkus.rest.jackson.optimization.enable-reflection-free-serializers=true
```

### Output of `uname -a` or `ver`

macOS

### Output of `java -version`

Java 21

### Quarkus version or git rev

Quarkus main

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/JacksonSerializerFactory.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/JacksonSerializerFactory.java
@@ -318,7 +318,7 @@ public class JacksonSerializerFactory {
 
         if (primitiveMethodName != null) {
             MethodDescriptor primitiveWriter = MethodDescriptor.ofMethod(JSON_GEN_CLASS_NAME, primitiveMethodName, "void",
-                    typeName);
+                    fieldSpecs.writtenType());
             bytecode.invokeVirtualMethod(primitiveWriter, jsonGenerator, arg);
             return;
         }
@@ -372,7 +372,7 @@ public class JacksonSerializerFactory {
 
     private String writeMethodForPrimitiveFields(String typeName) {
         return switch (typeName) {
-            case "java.lang.String" -> "writeString";
+            case "java.lang.String", "char", "java.lang.Character" -> "writeString";
             case "short", "java.lang.Short", "int", "java.lang.Integer", "long", "java.lang.Long", "float",
                     "java.lang.Float", "double", "java.lang.Double" ->
                 "writeNumber";
@@ -590,8 +590,37 @@ public class JacksonSerializerFactory {
         }
 
         ResultHandle toValueReaderHandle(BytecodeCreator bytecode, ResultHandle valueHandle) {
-            return methodInfo != null ? bytecode.invokeVirtualMethod(MethodDescriptor.of(methodInfo), valueHandle)
-                    : bytecode.readInstanceField(FieldDescriptor.of(fieldInfo), valueHandle);
+            ResultHandle handle = accessorHandle(bytecode, valueHandle);
+
+            handle = switch (fieldType.name().toString()) {
+                case "char", "java.lang.Character" -> bytecode.invokeStaticMethod(
+                        MethodDescriptor.ofMethod(Character.class, "toString", String.class, char.class), handle);
+                default -> handle;
+            };
+
+            return handle;
+        }
+
+        private ResultHandle accessorHandle(BytecodeCreator bytecode, ResultHandle valueHandle) {
+            if (methodInfo != null) {
+                if (methodInfo.declaringClass().isInterface()) {
+                    return bytecode.invokeInterfaceMethod(MethodDescriptor.of(methodInfo), valueHandle);
+                }
+                return bytecode.invokeVirtualMethod(MethodDescriptor.of(methodInfo), valueHandle);
+            }
+            return bytecode.readInstanceField(FieldDescriptor.of(fieldInfo), valueHandle);
+        }
+
+        String writtenType() {
+            return switch (fieldType.name().toString()) {
+                case "char", "java.lang.Character" -> "java.lang.String";
+                case "java.lang.Integer" -> "int";
+                case "java.lang.Short" -> "short";
+                case "java.lang.Long" -> "long";
+                case "java.lang.Double" -> "double";
+                case "java.lang.Float" -> "float";
+                default -> fieldType.name().toString();
+            };
         }
 
         private String[] rolesAllowed() {

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/Cat.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/Cat.java
@@ -14,4 +14,8 @@ public class Cat extends AbstractNamedPet {
     public void setPrivateAge(int privateAge) {
         this.privateAge = privateAge;
     }
+
+    public char getInitial() {
+        return getPublicName().charAt(0);
+    }
 }

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/ContainerDTO.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/ContainerDTO.java
@@ -1,0 +1,4 @@
+package io.quarkus.resteasy.reactive.jackson.deployment.test;
+
+public record ContainerDTO(NestedInterface nestedInterface) {
+}

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/NestedInterface.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/NestedInterface.java
@@ -1,0 +1,28 @@
+package io.quarkus.resteasy.reactive.jackson.deployment.test;
+
+public interface NestedInterface {
+
+    NestedInterface INSTANCE = new NestedInterface() {
+        @Override
+        public String getString() {
+            return "response";
+        }
+
+        @Override
+        public Integer getInt() {
+            return 42;
+        }
+
+        @Override
+        public char getCharacter() {
+            return 'a';
+        }
+    };
+
+    String getString();
+
+    Integer getInt();
+
+    char getCharacter();
+
+}

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonResource.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonResource.java
@@ -442,6 +442,12 @@ public class SimpleJsonResource extends SuperClass<Person> {
         return item.getContent().getName();
     }
 
+    @GET
+    @Path("/interface")
+    public ContainerDTO interfaceTest() {
+        return new ContainerDTO(NestedInterface.INSTANCE);
+    }
+
     public static class UnquotedFieldsPersonSerialization implements BiFunction<ObjectMapper, Type, ObjectWriter> {
 
         public static final AtomicInteger count = new AtomicInteger();

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonTest.java
@@ -34,7 +34,8 @@ public class SimpleJsonTest {
                                     NoopReaderInterceptor.class, TestIdentityProvider.class, TestIdentityController.class,
                                     AbstractPet.class, Dog.class, Cat.class, Veterinarian.class, AbstractNamedPet.class,
                                     AbstractUnsecuredPet.class, UnsecuredPet.class, SecuredPersonInterface.class, Frog.class,
-                                    Pond.class, FrogBodyParts.class, FrogBodyParts.BodyPart.class)
+                                    Pond.class, FrogBodyParts.class, FrogBodyParts.BodyPart.class, ContainerDTO.class,
+                                    NestedInterface.class)
                             .addAsResource(new StringAsset("admin-expression=admin\n" +
                                     "user-expression=user\n" +
                                     "birth-date-roles=alice,bob\n"), "application.properties");
@@ -507,6 +508,18 @@ public class SimpleJsonTest {
     }
 
     @Test
+    public void testInterface() {
+        RestAssured
+                .with()
+                .get("/simple/interface")
+                .then()
+                .statusCode(200)
+                .body("nestedInterface.int", Matchers.is(42))
+                .body("nestedInterface.character", Matchers.is("a"))
+                .body("nestedInterface.string", Matchers.is("response"));
+    }
+
+    @Test
     public void testSecureFieldOnAbstractClass() {
         // implementor with / without @SecureField returned
         testSecuredFieldOnAbstractClass("cat", "dog");
@@ -602,6 +615,7 @@ public class SimpleJsonTest {
                 .then()
                 .statusCode(200)
                 .body("publicName", Matchers.is("Garfield"))
+                .body("initial", Matchers.is("G"))
                 .body("privateName", Matchers.nullValue())
                 .body("veterinarian.name", Matchers.is("Dolittle"))
                 .body("veterinarian.title", Matchers.nullValue())
@@ -625,6 +639,7 @@ public class SimpleJsonTest {
                 .then()
                 .statusCode(200)
                 .body("publicName", Matchers.is("Garfield"))
+                .body("initial", Matchers.is("G"))
                 .body("privateName", Matchers.is("Monday"))
                 .body("privateAge", Matchers.is(4))
                 .body("veterinarian.name", Matchers.is("Dolittle"))

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonWithReflectionFreeSerializersTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonWithReflectionFreeSerializersTest.java
@@ -24,7 +24,8 @@ public class SimpleJsonWithReflectionFreeSerializersTest extends SimpleJsonTest 
                                     NoopReaderInterceptor.class, TestIdentityProvider.class, TestIdentityController.class,
                                     AbstractPet.class, Dog.class, Cat.class, Veterinarian.class, AbstractNamedPet.class,
                                     AbstractUnsecuredPet.class, UnsecuredPet.class, SecuredPersonInterface.class, Frog.class,
-                                    Pond.class, FrogBodyParts.class, FrogBodyParts.BodyPart.class)
+                                    Pond.class, FrogBodyParts.class, FrogBodyParts.BodyPart.class, ContainerDTO.class,
+                                    NestedInterface.class)
                             .addAsResource(new StringAsset("admin-expression=admin\n" +
                                     "user-expression=user\n" +
                                     "birth-date-roles=alice,bob\n" +

